### PR TITLE
[secrets] Fix + refactor bulk decryption.

### PR DESCRIPTION
**Pull request by pgavlin**
The bulk decryption logic in deployment deserialization did not recur
into slices or maps. This prevented the bulk decryption of nested secret
values. These changes fix that bug, improve test coverage, and refactor
the bulk decryption code for simplicity, clarity, and separation of
concerns.

Related to https://github.com/pulumi/home/issues/1842.

### CODE REVIEW

**Comment by pgavlin** [comment]
One note: bulk decryption will not see secrets nested _inside other secrets_. A secret is stored as the encrypted, JSON-serialized text of its plain value, so the contents of the plain value are not visible until the secret is decrypted. So the inner secrets in an example like this aren't visible to the scanner: `secret(secret(secret(....)))`. This should be fine, though, as those inner secrets are not encrypted; they're just marked as secrets (technically, they're encrypted using a no-op encrypter).

File: pkg/resource/stack/secrets.go [hunk 1]
```diff
@@ -154,3 +150,35 @@ func (c *cachingCrypter) encryptSecret(secret *resource.Secret, plaintext string
 func (c *cachingCrypter) insert(secret *resource.Secret, plaintext, ciphertext string) {
 	c.cache[secret] = cacheEntry{plaintext, ciphertext}
 }
+
+// mapDecrypter is a Decrypter with a preloaded cache. This decrypter is used specifically for deserialization,
+// where the deserializer is expected to prime the cache by scanning each resource for secrets, then decrypting all
+// of the discovered secrets en masse. Although each call to Decrypt _should_ hit the cache, a mapDecrypter does
+// carry an underlying Decrypter in the event that a secret was missed.
+type mapDecrypter struct {
```
Comment (Frassle) [inline pkg/resource/stack/secrets.go:158]: Feels like this is similar to the cachingCrypter idea. Wonder if all the caching logic could just be done once, although this caches by cipertext and cachingCrypter is by Secret pointers so maybe makes no sense to combine.

File: sdk/go/common/resource/config/crypt.go [hunk 2]
```diff
@@ -36,6 +36,10 @@ type Encrypter interface {
 // Decrypter decrypts encrypted ciphertext to its plaintext representation.
 type Decrypter interface {
 	DecryptValue(ciphertext string) (string, error)
+}
+
+// BulkDecrypter is a Decrypter that also supports bulk decryption of secrets.
+type BulkDecrypter interface {
```
[the remark this answers is not in the document]
Comment (pgavlin) [reply sdk/go/common/resource/config/crypt.go:42]: Ah, yep--good catch.

File: pkg/resource/stack/secrets.go (hunk 1 above)
Comment (pgavlin) [reply pkg/resource/stack/secrets.go:158]: > this caches by cipertext and cachingCrypter is by Secret pointers so maybe makes no sense to combine.

This is what led me to keep the two separate. My first instinct was to combine them, but the differences in key type and semantics led me down the proposed approach.

Comment (pgavlin) [reply pkg/resource/stack/secrets.go:164]: Added a comment re: the distinction

File: pkg/resource/stack/deployment.go [hunk 3]
```diff
@@ -456,34 +456,24 @@ func SerializePropertyValue(prop resource.PropertyValue, enc config.Encrypter,
...
+			collectCiphertexts(ciphertexts, v)
+		}
+	case map[string]interface{}:
+		if prop[resource.SigKey] == resource.SecretSig {
+			if ciphertext, cipherOk := prop["ciphertext"].(string); cipherOk {
+				*ciphertexts = append(*ciphertexts, ciphertext)
+			}
+		} else {
+			for _, v := range prop {
+				collectCiphertexts(ciphertexts, v)
```
Comment (dixler) [inline pkg/resource/stack/deployment.go:473]: nice refactor

**Review by Frassle — APPROVED** [review]

**Pull request merged**